### DISCUSSION TITLE
Custom assets on server side only

### DISCUSSION THREAD

**Discussion by maxdzin** [nuxt/nuxt#20790] [Questions] [answered]
Hello Nuxters,

I want to know if it is possible to keep some assets files (after build) that could be accessible on the server side only.
Maybe in `server/assets` dir, but that's not supported.
There are `public` and `server` (with `chunks` and `node_modules`) dirs in the `.output` after the build.

And I need the possibility to keep, let's say, a few JSON files those doesn't need to appear in `public` directory (so no access from the client side) but can be read by server means.

So, maybe someone has an idea on how to achieve that?

**Comment by maxdzin** [comment]
**UPD**:

I found that if I store some file `server/assets/test.json`:
```json
{
  "test": "value"
}
```
 after the build it appears as `.output/server/chunks/raw/test.mjs`:
 ```mjs
// ROLLUP_NO_REPLACE 
 const test = "{\n  \"test\": \"value\"\n}\n";

export { test as default };
//# sourceMappingURL=test.mjs.map
```

Will try to dig more in that direction.

**Comment by maou-shonen** [comment]
[nitro custom-server-assets](https://nitro.unjs.io/guide/assets#custom-server-assets)?

**Reply by maxdzin** [reply]
Will check, thank you!

**Reply by maxdzin** [reply]
@maou-shonen OK, so that's what I get:
- Config:
```ts
  nitro: {
    serverAssets: [
      {
        baseName: '',
        dir: 'assets', // I tried different variations here
      },
    ],
  },
...
```

- Files:
`server\assets\templates\test.txt`

- Code:
```ts
...
  const template = await useStorage().getItem('root:server:assets:templates:default.txt')
```

And while in dev, it works well, but in production there are still no assets dir.

**Reply by maou-shonen** [reply]
Oh...I understand what you want is an extra public directory, and not compiled by rollup.

you should be able to mount a [Storage Layer](https://nitro.unjs.io/guide/storage)
```ts
export default defineNuxtConfig({
  nitro: {
    storage: {
      serverAssets: {
        driver: 'fs',
        base: '.output/server/data'
      }
    },
    devStorage: {
      serverAssets: {
        driver: 'fs',
        base: './data'  // in nuxt3 -> ~/server/data
      }
    }
  }
})
```

Then copy this directory after build
```json
// package.json
{
 "scripts": {
    "build": "nuxt build && cp -r ./server/data .output/server/data", 
  },
}
```

**Reply by maxdzin** [reply, marked as answer]
Hi @maou-shonen,
Thank you! It's not really necessary.
I found out how to do it.
So, thank you for pointing to the Nitro documentation.

The directory `server/assets/mail/...`.

Config:
```ts
nitro: {
    serverAssets: [
      {
        baseName: 'mail',
        dir: './assets/mail',
      },
    ],
  },
```

In code:
```ts
const layoutSource = await useStorage('assets:mail').getItem('layouts:default.hbs')
```

I hope that helps someone as well.

**Reply by ReaganM02** [reply]
Thank you for sharing your solution. I had the same issue with email templates too in production 😄

**Reply by bart** [reply]
Hey @maxdzin I had the same question about a week ago and came up with the same solution but instead of an hbs file I need image files (PNGs) on the server-side. Problem now is that the `useStorage().getItem()` call gets the content of given file in UTF8 which is a problem for image files. getItemRaw() should return the Buffer instead but it doesn't seem to work because `useStorage()` in combination with `serverAssets` configuration uses the memory driver of the unstorage package. Does anyone have an idea how to solve this? Thanks in advance!
